libbpf-rs/src/tc.rs

```diff
@@ -0,0 +1,291 @@
+use nix::errno;
+use nix::errno::Errno::{EEXIST, EINVAL, ENOENT};
+
+use crate::*;
+
+pub type TcAttachPoint = libbpf_sys::bpf_tc_attach_point;
+pub const TC_INGRESS: TcAttachPoint = libbpf_sys::BPF_TC_INGRESS;
+pub const TC_EGRESS: TcAttachPoint = libbpf_sys::BPF_TC_EGRESS;
+pub const TC_CUSTOM: TcAttachPoint = libbpf_sys::BPF_TC_CUSTOM;
+
+pub type TcFlags = libbpf_sys::bpf_tc_flags;
+pub const BPF_TC_F_REPLACE: TcFlags = libbpf_sys::BPF_TC_F_REPLACE;
+
+// from kernel @ include/uapi/linux/pkt_sched.h
+pub const TC_H_INGRESS: u32 = 0xFFFFFFF1;
+pub const TC_H_CLSACT: u32 = TC_H_INGRESS;
+pub const TC_H_MIN_INGRESS: u32 = 0xFFF2;
+pub const TC_H_MIN_EGRESS: u32 = 0xFFF3;
+pub const TC_H_MAJ_MASK: u32 = 0xFFFF0000;
+pub const TC_H_MIN_MASK: u32 = 0x0000FFFF;
+
+/// The BPF TC subsystem has different control paths from other BPF programs
+/// As such a BPF program using a TC Hook (SEC("classifier") | SEC("tc")) must be operated
+/// more independently from other [libbpf-rs::Program]s where the program itself can
+/// be attach_*() a program using TC must grab a hook point, or create one
+/// and attach from the TcHook
+///
+/// Documentation about the libbpf TC interface can be found here
+/// https://lwn.net/ml/bpf/20210512103451.989420-3-memxor@gmail.com/
+///
+/// An example of using a BPF TC program can be seen in [examples/tc_whitelist_ports/src/main.rs]
+///
+/// Represents a Hook Point for a Traffic Control (TC) bpf program
+///
+/// This object exposes operations to create and attach a bpf_tc_hook
+/// into the TC subsystem
+#[derive(Clone, Copy, Debug)]
+pub struct TcHook {
+    hook: libbpf_sys::bpf_tc_hook,
+    opts: libbpf_sys::bpf_tc_opts,
+}
+
+impl TcHook {
+    /// Create a New TcHook given the file descriptor of the loaded SEC("tc") Program
+    /// See: [libbpf-rs::Program]
+    pub fn new(fd: i32) -> Self {
+        let mut tc_hook = TcHook {
+            hook: libbpf_sys::bpf_tc_hook::default(),
+            opts: libbpf_sys::bpf_tc_opts::default(),
+        };
+
+        tc_hook.hook.sz = std::mem::size_of::<libbpf_sys::bpf_tc_hook>() as u64;
+        tc_hook.opts.sz = std::mem::size_of::<libbpf_sys::bpf_tc_opts>() as u64;
+        tc_hook.opts.prog_fd = fd;
+
+        tc_hook
+    }
+
+    /// create a new TcHook
+    ///
+    /// if a TcHook already exists with the same parameters as the hook calling
+    /// create(), this function will still succeed.
+    ///
+    /// Will always fail on a TC_CUSTOM hook
+    pub fn create(&mut self) -> Result<Self> {
+        let err = unsafe { libbpf_sys::bpf_tc_hook_create(&mut self.hook as *mut _) };
+        // the hook may already exist, this is not an error
+        if err != 0 && err != -(EEXIST as i32) {
```

Comment:
> I'm not seeing where bpf_tc_hook_create can return -EEXIST

This error is coming out of the kernel rather than libbpf.  I shall track it down.

I **think** the error is coming out from here (https://elixir.bootlin.com/linux/latest/source/net/sched/cls_api.c#L2111).  But this is a wild guess.  During development I noticed the error during a double create, and tried to track through libbpf to see where the EEXIST comes from, and stopped as I could not track it via libbpf code spelunking.

Reply:
I think it's coming from here https://elixir.bootlin.com/linux/latest/source/net/sched/sch_api.c#L1625 - `bpf_tc_hook_create` triggers the `tc_modify_qdisc` function in `sch_api.c` since we send  a `RTM_NEWQDISC` message when we create the tc hook (in tools/lib/bpf/netlink.c)